Add Fermi surface component

utf opened the pull request:
Fixes #258.

Initial implementation. A simple component and app with no controls. I still need to add tests.

Interested in feedback for what you'd like to see at this stage @mkhorton.

In the documentation it says: 
>For interactivity, implement the generate_callbacks(app, cache) method. Inside this method you can define callbacks to be associated with the component. At minimum, make sure that if the component’s main dcc.Store is updated with a new object (for example, a new crystal structure is set for StructureMoleculeComponent) that the state of the component is also updated appropriately.

I'm not sure how to do that. Are there any examples?

<img width="1026" alt="Screenshot 2022-10-14 at 19 38 29" src="https://user-images.githubusercontent.com/1330638/195917648-d4a56f06-4c55-494a-ad98-183ac81ea92e.png">

mkhorton commented:
Thanks for the PR @utf!

> I'm not sure how to do that. Are there any examples?

Yes, absolutely. You might want to refactor so it looks more like this:

```

    @staticmethod
    def get_figure(fermi_surface: FermiSurface) -> go.Figure():
        # code moved from layout()
        ...

     def layout(self):
         """Get a Dash layout for the component."""

         if initial_data := self.initial_data["default"]:
              figure = self.get_figure(initial_data)
        else:
              figure = None

         return html.Div(
             [
                 Loading(
                     [
                         dcc.Graph(
                             id=self.id("fermi-surface-graph")  # note this `id` was added to allow for callbacks
                             figure=figure,
                             config={"displayModeBar": False},
                             responsive=True,
                         )
                     ],
                     id=self.id("fermi-surface-div"),
                 )
             ]
         )


     def generate_callbacks(self, app, cache):

         @app.callback(
            Output(self.id("fermi-surface-graph"), "figure"),
            Input(self.id(), "data")
        )
        def update_plot(fermi_surface):
              # if update_plot is slow, an @cache decorator can be added here
              fermi_surface = self.get_data(fermi_surface)  # converts back to object
              return self.get_plot(fermi_surface)
```


In your example, I would also think about what interactivity might be useful to the user:

1. Are there any input options you would like to include, so that the generation of the plot might be modified? The answer here might be no -- the only kwarg I see is `symprec` which perhaps is not useful for the end user.
2. Is there any interactivity you would like to have with the Fermi surface, e.g. to click on a specific surface or a specific point and report additional information? 

For both 1. and 2. I can outline specific ways of doing this, but only if there was some functionality you wanted to add.

For an initial PR, this is sufficient even without additional functionality and looks really nice!

janosh commented on crystal_toolkit/apps/examples/fermi_surface.py:
```diff
@@ -0,0 +1,38 @@
+import dash
+
+# dos and bs data from local jsons
+from monty.serialization import loadfn
```
Iirc, `monty` is just a transient dep atm. Should we add it to `pyproject.toml` since it's used in a dozen other places?

mkhorton replied:
Honestly I don't think it makes much odds -- it's so baked into pymatgen, and pymatgen is such a core requirement, that I think we'd only explicitly need to list it if we had a different minimum version. But by all means add if you feel strongly!

utf commented:
Thanks Matt. That makes sense about the store. I was going off the band structure component which doesn't seem to implement that sort of workflow but maybe I missed it.

For controls I was thinking:

- Show/hide unit cell edges.
- Show/hide high-symmetry labels.
- Show/hide Fermi surface properties (if they've been calculated) such as group velocity.
- Controls for colormap for FS properties.

At the moment I don't see any easy and useful interactivity elements. One thing for the future would be clicking an isosurface and seeing that band highlighted in the band structure, in addition to a summary of the properties for that surface (surface area, dimensionality, property average). But that is a job for another PR IMO.

mkhorton commented:
> For controls I was thinking:
> 
> Show/hide unit cell edges.
> Show/hide high-symmetry labels.
> Show/hide Fermi surface properties (if they've been calculated) such as group velocity.
> Controls for colormap for FS properties

If you want to add a control, the pattern we've been using are the methods:

```
numerical_input = self.get_numerical_input(kwarg_label="...", default="...", label="...", help_str="...")
slider_input = self.get_slider_input(...)
# also bool_input, choice_input, dict_input, ...
```

which can then be added to your layout.

These interfaces for adding controls are used for two reasons, (1) it allows us to add `label`, `help_str` etc. in a standardized way, (2) the input id are set such that we can retrieve the values of these controls in a standardized way, as shown below.

To retrieve the values inside your callback, you would add an additional input:

```
Input(self.get_all_kwargs_id(), "value")
```

and then in the body of the callback:

```
kwargs = self.reconstruct_kwargs_from_state()
```

and this returns a dictionary of the values of the controls specified.

I should add this to the docs, but we were still establishing the pattern.

utf commented:
Hi @mkhorton, I've added some controls, including the ability to select the colour scheme for Fermi surfaces with projections. See the example screenshots for an example. The colour scheme selection is only enabled if the Fermi surface has projections to begin with. I've also added a test.

Unfortunately, CI is failing due to installing boltztrap. This is unrelated to this PR as I didn't change any of the dependencies or CI at all. It seems to be something to do with poetry installing boltztrap through pip and missing a the "packing" package. But I'm not familiar enough with poetry to know how to fix it.

From my point of view, this should be ready for final review/to merge.

<img width="1184" alt="Screenshot 2022-10-17 at 14 41 41" src="https://user-images.githubusercontent.com/1330638/196194941-d66afd63-421f-45d0-99b4-1719f4c0c435.png">

<img width="1184" alt="Screenshot 2022-10-17 at 14 41 48" src="https://user-images.githubusercontent.com/1330638/196194971-4c339862-5067-4d69-bc46-0a5e9c4e0b3f.png">

janosh commented:
I think a similar Boltztrap2 error [popped up over at `pymatgen`](https://github.com/materialsproject/pymatgen/actions/runs/3315144174/jobs/5475347895#step:4:788) during `pip install BoltzTraP2` (i.e. no `poetry`). But there it seems to only occur on Windows.

@utf I saw this thread from 2021 which also seems related: https://github.com/fermisurfaces/IFermi/issues/17#issuecomment-796936649
Couldn't find your issue on the [Boltztrap2 repo](https://gitlab.com/sousaw/BoltzTraP2). Do you remember if they offered any advice?

utf commented:
I don't think https://github.com/fermisurfaces/IFermi/issues/17#issuecomment-796936649  was related. This is a new issue which is only present with the latest version of boltztrap. It seems like packaging is required to be installed before installing boltztrap2.

mkhorton commented:
Hi @utf, PR is approved -- please feel welcome to merge this, and to merge subsequent PRs if you want to make additional changes.

I'm giving all collaborators (e.g. paper co-authors) the ability to merge their own PRs in the hopes of making the project feel more collaborative, and I'll still do a final check of anything merged prior to release as project maintainer.

utf commented:
Great, thanks Matt!
